libbpf-rs/src/tc.rs

```diff
@@ -0,0 +1,284 @@
+pub const TC_H_MAJ_MASK: u32 = 0xFFFF0000;
+pub const TC_H_MIN_MASK: u32 = 0x0000FFFF;
+
+/// The BPF TC subsystem has different control paths from other BPF programs
+/// As such a BPF program using a TC Hook (SEC("classifier") | SEC("tc")) must be operated
+/// more independently from other [libbpf-rs::Program]s.
+///
+/// Documentation about the libbpf TC interface can be found here
+/// https://lwn.net/ml/bpf/20210512103451.989420-3-memxor@gmail.com/
+///
+/// An example of using a BPF TC program can be seen in [examples/tc_whitelist_ports/src/main.rs]
+///
+/// Represents a Hook Point for a Traffic Control (TC) bpf program
+///
+/// The TcHook struct exposes operations to create, attach, query and destroy
+/// a bpf_tc_hook using the TC subsystem
+#[derive(Clone, Copy, Debug)]
+pub struct TcHook {
+    hook: libbpf_sys::bpf_tc_hook,
+    opts: libbpf_sys::bpf_tc_opts,
+}
+
+impl TcHook {
+    /// Create a New TcHook given the file descriptor of the loaded SEC("tc") Program
+    /// See: [libbpf-rs::Program]
+    pub fn new(fd: i32) -> Self {
+        let mut tc_hook = TcHook {
+            hook: libbpf_sys::bpf_tc_hook::default(),
+            opts: libbpf_sys::bpf_tc_opts::default(),
+        };
+
+        tc_hook.hook.sz = std::mem::size_of::<libbpf_sys::bpf_tc_hook>() as u64;
+        tc_hook.opts.sz = std::mem::size_of::<libbpf_sys::bpf_tc_opts>() as u64;
+        tc_hook.opts.prog_fd = fd;
+
+        tc_hook
+    }
+
+    /// create a new TcHook
+    ///
+    /// if a TcHook already exists with the same parameters as the hook calling
+    /// create(), this function will still succeed.
+    ///
+    /// Will always fail on a TC_CUSTOM hook
+    pub fn create(&mut self) -> Result<Self> {
+        let err = unsafe { libbpf_sys::bpf_tc_hook_create(&mut self.hook as *mut _) };
+        // the hook may already exist, this is not an error
+        if err != 0 && err != -(EEXIST as i32) {
+            Err(Error::System(err))
+        } else {
+            Ok(*self)
+        }
+    }
+
+    /// Set the interface to be used
+    ///
+    /// Interfaces can be listed by using ip link command from the iproute2 software package
+    pub fn ifindex(&mut self, idx: i32) -> &mut Self {
+        self.hook.ifindex = idx;
+        self
+    }
+
+    /// Set what type of TC point to attach onto
+    ///
+    /// TC_EGRESS, TC_INGRESS, or TC_CUSTOM
+    ///
+    /// An TC_EGRESS|TC_INGRESS hook can be used as an attach_point for calling
+    /// the destroy() method to remove the clsact bpf tc qdisc, but cannot
+    /// be used to attach()
+    pub fn attach_point(&mut self, ap: TcAttachPoint) -> &mut Self {
+        self.hook.attach_point = ap;
+        self
+    }
+
+    /// Set the parent of a hook
+    /// Will return EINVAL if set upon an TC_EGRESS/TC_INGRESS/TC_EGRESS|TC_INGRESS hook
+    ///
+    /// Must be set on a TC_CUSTOM hook
+    /// Current acceptable values are TC_H_CLSACT for maj, and
+    /// TC_H_MIN_EGRESS or TC_H_MIN_INGRESS for min
```

Comment:
I don't see where we enforce this in `pub fn parent`. Is this documentation meant to be placed here?

Reply:
The documentation is unclear.  If a parent is set upon a non-TC_CUSTOM hook, the attach() call will return EINVAL, as setting the parent is only valid for TC_CUSTOM.

I've updated the documentation to be more clear.